ntoskrnl/mm/ARM3/pagfault.c

```diff
@@ -1676,608 +1762,352 @@ MiDispatchFault(IN ULONG FaultCode,
-            MiUnlockProcessWorkingSet(CurrentProcess, CurrentThread);
+    /* Bring PDE in */
+    for (ULONG level = _MI_PAGING_LEVELS; level >= 2; level--)
+    {
+        PMMPDE PointerPde;
+        if (level == 4)
+            PointerPde = MiAddressToPde(MiAddressToPde(Address));
+        else if (level == 3)
+            PointerPde = MiAddressToPde(MiAddressToPte(Address));
+        else
+            PointerPde = MiAddressToPde(Address);
+        Status = MiBringInPageTable(FaultCode,
+                                    PointerPde,
+                                    Address,
+                                    CurrentProcess,
+                                    level,
+                                    &ProtectionCode,
+                                    &Vad);
+        if (!NT_SUCCESS(Status))
             return Status;
-        }
+    }
 
-        /* Resolve a demand zero fault */
-        MiResolveDemandZeroFault(PointerPte,
-                                 PointerPde,
-                                 MM_EXECUTE_READWRITE,
-                                 CurrentProcess,
-                                 MM_NOIRQL);
-#if _MI_PAGING_LEVELS >= 3
-        MiIncrementPageTableReferences(PointerPte);
-#endif
+    /* Quick path for Page tables */
+    if (MiIsUserPageTable(Address))
+    {
+        /* Only valid from KernelMode */
+        ASSERT(Mode == KernelMode);
 
-#if MI_TRACE_PFNS
-        UserPdeFault = FALSE;
-        /* Update debug info */
-        if (TrapInformation)
-            MiGetPfnEntry(PointerPde->u.Hard.PageFrameNumber)->CallSite = (PVOID)((PKTRAP_FRAME)TrapInformation)->Eip;
-        else
-            MiGetPfnEntry(PointerPde->u.Hard.PageFrameNumber)->CallSite = _ReturnAddress();
-#endif
-        /* We should come back with APCs enabled, and with a valid PDE */
-        ASSERT(KeAreAllApcsDisabled() == TRUE);
-        ASSERT(PointerPde->u.Hard.Valid == 1);
+        return MiBringInPageTable(FaultCode,
+                                  MiPteToPde(Address),
+                                  Address,
+                                  CurrentProcess,
+                                  _MI_PAGING_LEVELS, /* Let caller handle references */
+                                  &ProtectionCode,
+                                  &Vad);
     }
-    else
+
+    /* Make sure we're at the right place */
+    ASSERT((Address < MmSystemRangeStart) || MI_IS_HYPER_ADDRESS(Address));
+
+    /* Hyper space is not paged out yet */
+    if (MI_IS_HYPER_ADDRESS(Address))
     {
-        /* Not yet implemented in ReactOS */
-        ASSERT(MI_IS_PAGE_LARGE(PointerPde) == FALSE);
+        UNIMPLEMENTED_DBGBREAK("No support for hyper space in page fault handler!\n");
+        return STATUS_ACCESS_VIOLATION;
     }
 
-    /* Now capture the PTE. */
-    TempPte = *PointerPte;
+    /* PDE was just brought in. Barrier. */
+    KeMemoryBarrierWithoutFence();
```

Comment:
Shouldn't be needed

Reply:
I'm not sure. MiBringInPageTable is not forced inline.